tdx/mrtd_from_ovmf.go

```diff
@@ -0,0 +1,79 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tdx
+
+import (
+	"fmt"
+
+	"github.com/google/gce-tcb-verifier/ovmf"
+)
+
+const (
+	gib      = 0x40000000
+	threeGiB = 0xc0000000
+)
+
+func regionsForGB(sizeInGB int) []ovmf.GuestPhysicalRegion {
+	return []ovmf.GuestPhysicalRegion{
+		{Start: 0, Length: threeGiB},                           // Lower 3GiB
+		{Start: 0xffe00000, Length: 0x200000},                  // Top 2MiB under 4GiB for TDVF
+		{Start: 0x100000000, Length: uint64(sizeInGB-3) * gib}, // Rest of memory
```

Comment:
If the first 2 regions are for the first 4GiB of memory, shouldn't this be (sizeInGB-4) * gib? And what if sizeInGB < 4?

Reply:
There's an "MMIO hole" between 3GiB and 4GiB, except for the firmware itself, so you've got 0-3gib, 4gib-2mib-4gib, and then 4gib-(rest of memory) which has rounded up to the nearest gib. I've confirmed this with logs out of vanadium for the memory banks it uses.